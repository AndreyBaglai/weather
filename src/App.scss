.container {
  max-width: 1920px;
  padding: 0 40px;
}

.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 80px;

  .form {
    display: flex;
    align-items: center;

    .city-field {
      width: 250px;
      height: 40px;
      background: #ffffff;
      box-shadow: 0px 3px 6px #00000029;
      border-radius: 5px;
      font-size: 24px;
      padding-left: 6px;
      border: none;
      margin-right: 20px;
    }

    #addBtn {
      width: 112px;
      height: 40px;
      background: #459de9;
      box-shadow: 0px 3px 6px #00000029;
      border-radius: 5px;
      border: none;
      color: #ffffff;
      text-transform: capitalize;
      font-size: 18px;
      cursor: pointer;
    }
  }

  .select-lang {
    width: 71px;
    height: 24px;
    background: #ffffff;
    box-shadow: 0px 3px 6px #00000029;
    font-size: 16px;
    border: none;
  }
}

.main {
  .city-weather {
    display: flex;
    justify-content: left;
    flex-wrap: wrap;

    .day {
      margin: 20px 40px 40px 0;
      width: 350px;
      height: 257px;
      background: #fff1fe;
      box-shadow: 0px 3px 6px #00000029;
      border-radius: 5px;
      padding: 14px;
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      &-top, &-bottom {
        display: flex;
        justify-content: space-between;
      }

      &-bottom {
        align-items: flex-end;
      }

      .graphic {
        width: 320px;
        height: 51px;
        background: transparent linear-gradient(180deg, #ff715b 0%, #fff4f4 100%);
        opacity: 0.3;
      }

      .temperature-wrapper {
        position: relative;
        margin-bottom: 20px;

        .temperature {
          font-size: 48px;
          font-weight: bold;
          margin-right: 10px;
        }

        .metric {
          font-weight: normal;
          position: relative;
          bottom: 24px;
        }
      }

      .weather-info {
        text-align: right;
        p {
          margin: 4px 0;

          .value {
            color: #459de9;
          }
        }
      }
    }
  }
}
