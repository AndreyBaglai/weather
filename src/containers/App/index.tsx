import React, { useEffect } from 'react';
import { observer } from 'mobx-react-lite';

import Header from 'components/Header';
import Main from 'components/Main';

import { useStore } from 'stores';

import { getCardsFromLS, getLangFromLS, setLangToLS } from '../../services/localStorage';

import styles from './styles.module.scss';

const App = observer(() => {
  const { cardsStore, languageStore } = useStore();
  
  useEffect(() => {
    const currentLang = getLangFromLS();
    const currentCards = getCardsFromLS();

    if (currentCards.length) {
      cardsStore.updateCards(getCardsFromLS());
    }

    currentLang !== '' ? languageStore.changeLang(currentLang) : setLangToLS(languageStore.lang);
  }, []);

  return (
    <div className={styles.container}>
      <Header />
      {cardsStore.cards.length === 0 ? <p>Please, input city name</p> : <Main />}
    </div>
  );
});

export default App;
