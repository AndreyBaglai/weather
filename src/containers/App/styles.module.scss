.container {
  max-width: 1920px;
  padding: 0 40px;
  box-sizing: border-box;
  margin: 0 auto;

  .text {
    text-align: center;
  }
}

@media screen and (max-width: 700px) {
  .container {
    padding: 0 10px;
  }
}