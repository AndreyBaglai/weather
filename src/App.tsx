import React, { useEffect, useState } from 'react';
import './App.scss';
import Header from './components/Header/Header';
import Main from './components/Main/Main';
import { CardModel } from './model/card-model';
import { getCardsFromLS, setCardsToLS } from './services/localStorage';
import { getWeatherByCity } from './services/weather-api';

// по городу https://api.openweathermap.org/data/2.5/weather?q=Kiev&units=metric&lang=en&appid=0f3e903b21bbba52b9410fe0033434f1 берем координаты и по координатам
// https://api.openweathermap.org/data/2.5/onecall?lat=50.450001&lon=30.523333&exclude=hourly,minutely&lang=en&units=metric&appid=0f3e903b21bbba52b9410fe0033434f1

function App() {
  const [cards, setCards] = useState<CardModel[]>([]);
  const [isCelsius, setIsCelsius] = useState(true);

  const init = () => {
    const cards = getCardsFromLS();

    if (cards.length) {
      setCards(cards);
    }
  };

  useEffect(() => {
    init();
  }, []);

  const onSubmitForm = async (e: React.FormEvent) => {
    e.preventDefault();
    const target = e.target as HTMLFormElement;
    const input = target.querySelector('#city') as HTMLInputElement;

    try {
      if (input.value === '') return;

      const data = await getWeatherByCity(input.value);
      if (!data) return;

      const card: CardModel = {
        city: data.name,
        time: data.dt,
        country: data.sys.country,
        temperature: data.main.temp,
        humidity: data.main.humidity,
        pressure: data.main.pressure,
        feels: data.main.feels_like,
        icon: data.weather[0].icon,
        text_icon: data.weather[0].main,
        wind_speed: data.wind.speed,
      };
      input.value = '';
      setCards([card, ...cards]);
      setCardsToLS([card, ...cards]);
    } catch (err) {
      input.value = '';
      console.log('Incorrect city name');
    }
  };

  const onChangeTemperature = (e: React.MouseEvent) => {
    const currentTarget = e.currentTarget as HTMLElement;
    const target = e.target as HTMLElement;

    if (
      (target.classList.contains('fahrenheit') && !isCelsius) ||
      (target.classList.contains('celsius') && isCelsius)
    ) {
      return;
    }

    const temperatureEl = currentTarget.querySelector('.temperature') as HTMLElement;
    const temperature = Number(
      isCelsius ? temperatureEl?.textContent?.slice(1) : temperatureEl?.textContent,
    );

    if (target.classList.contains('fahrenheit') && temperatureEl) {
      temperatureEl.textContent = String(Math.ceil((temperature * 9) / 5 + 32));
      setIsCelsius(false);
    } else {
      const temp = Math.floor((5 / 9) * (temperature - 32));
      temperatureEl.textContent = String(temp > 0 ? `+${temp}` : `-${temp}`);
      setIsCelsius(true);
    }
  };

  return (
    <div className="container">
      <Header onSubmitForm={onSubmitForm} />
      {cards.length === 0 ? (
        <p>Please, input city name</p>
      ) : (
        <Main onChangeTemperature={onChangeTemperature} cards={cards} />
      )}
    </div>
  );
}

export default App;
