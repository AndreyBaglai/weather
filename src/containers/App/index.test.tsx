import React from 'react';

import { render, screen } from '@testing-library/react';

import App from './index';

describe('App', () => {
  it('checking the structure of the application', () => {
    render(<App />);
    // header
    expect(screen.getByRole('banner')).toBeInTheDocument;

    // input
    expect(screen.getByRole('textbox')).toBeInTheDocument;

    //buttons add and remove
    expect(screen.getAllByRole('button')).toBeInTheDocument;

    //select
    expect(screen.getByRole('combobox')).toBeInTheDocument;

    //options
    expect(screen.getAllByRole('option')).toBeInTheDocument;

    //h2
    expect(screen.getByRole('heading')).toBeInTheDocument;
  });
});

/*
Search variants:
  getBy:                    queryby:                    findBy:
- getByText               - queryByText               - findByText
- getByRole               - queryByRole               - findByRole
- getByLabelText          - queryByLabelText          - findByLabelText
- getByPlaceholderText    - queryByPlaceholderText    - findByPlaceholderText
- getByAltText            - queryByAltText            - findByAltText
- getByDisplayValue       - queryByDisplayValue       - findByDisplayValue
- getAllBy                - queryAllBy                - findAllBy
*/

/*
Assertive Functions:
- toBeDisabled            - toBeEnabled               - toBeEmpty
- toBeEmptyDOMElement     - toBeInTheDocument         - toBeInvalid
- toBeRequired            - toBeValid                 - toBeVisible
- toContainElement        - toContainHTML             - toHaveAttribute
- toHaveClass             - toHaveFocus               - toHaveFormValues
- toHaveStyle             - toHaveTextContent         - toHaveValue
- toHaveDisplayValue      - toBeChecked               - toBePartiallyChecked
- toHaveDescription
*/
