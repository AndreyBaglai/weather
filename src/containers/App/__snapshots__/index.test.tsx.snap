// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render App component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <header
      class="header"
    >
      <form
        class="form"
      >
        <input
          class="cityField"
          id="city"
          name="city"
          placeholder="placeholder"
          type="text"
        />
        <button
          class="addBtn"
        >
          buttons.add
        </button>
      </form>
      <button
        class="removeAll"
      >
        buttons.remove
      </button>
      <select
        class="selectLang"
        id="lang"
        name="language"
      >
        <option
          value="en"
        >
          EN
        </option>
        <option
          value="ru"
        >
          RU
        </option>
        <option
          value="uk"
        >
          UK
        </option>
        <option
          value="he"
        >
          HE
        </option>
      </select>
    </header>
    <h2
      class="text"
    >
      empty
    </h2>
  </div>
</DocumentFragment>
`;

exports[`App render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="container-element"
      >
        <header
          class="header"
          data-testid="header-element"
        >
          <form
            class="form"
          >
            <input
              class="cityField"
              id="city"
              name="city"
              placeholder="placeholder"
              type="text"
            />
            <button
              class="addBtn"
            >
              buttons.add
            </button>
          </form>
          <button
            class="removeAll"
          >
            buttons.remove
          </button>
          <select
            class="selectLang"
            id="lang"
            name="language"
          >
            <option
              value="en"
            >
              EN
            </option>
            <option
              value="ru"
            >
              RU
            </option>
            <option
              value="uk"
            >
              UK
            </option>
            <option
              value="he"
            >
              HE
            </option>
          </select>
        </header>
        <h2
          class="text"
        >
          empty
        </h2>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="container-element"
    >
      <header
        class="header"
        data-testid="header-element"
      >
        <form
          class="form"
        >
          <input
            class="cityField"
            id="city"
            name="city"
            placeholder="placeholder"
            type="text"
          />
          <button
            class="addBtn"
          >
            buttons.add
          </button>
        </form>
        <button
          class="removeAll"
        >
          buttons.remove
        </button>
        <select
          class="selectLang"
          id="lang"
          name="language"
        >
          <option
            value="en"
          >
            EN
          </option>
          <option
            value="ru"
          >
            RU
          </option>
          <option
            value="uk"
          >
            UK
          </option>
          <option
            value="he"
          >
            HE
          </option>
        </select>
      </header>
      <h2
        class="text"
      >
        empty
      </h2>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App take a snapshot when  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="container-element"
      >
        <header
          class="header"
          data-testid="header-element"
        >
          <form
            class="form"
          >
            <input
              class="cityField"
              id="city"
              name="city"
              placeholder="placeholder"
              type="text"
            />
            <button
              class="addBtn"
            >
              buttons.add
            </button>
          </form>
          <button
            class="removeAll"
          >
            buttons.remove
          </button>
          <select
            class="selectLang"
            id="lang"
            name="language"
          >
            <option
              value="en"
            >
              EN
            </option>
            <option
              value="ru"
            >
              RU
            </option>
            <option
              value="uk"
            >
              UK
            </option>
            <option
              value="he"
            >
              HE
            </option>
          </select>
        </header>
        <h2
          class="text"
        >
          empty
        </h2>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="container-element"
    >
      <header
        class="header"
        data-testid="header-element"
      >
        <form
          class="form"
        >
          <input
            class="cityField"
            id="city"
            name="city"
            placeholder="placeholder"
            type="text"
          />
          <button
            class="addBtn"
          >
            buttons.add
          </button>
        </form>
        <button
          class="removeAll"
        >
          buttons.remove
        </button>
        <select
          class="selectLang"
          id="lang"
          name="language"
        >
          <option
            value="en"
          >
            EN
          </option>
          <option
            value="ru"
          >
            RU
          </option>
          <option
            value="uk"
          >
            UK
          </option>
          <option
            value="he"
          >
            HE
          </option>
        </select>
      </header>
      <h2
        class="text"
      >
        empty
      </h2>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App take a snapshot when render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="container-element"
      >
        <header
          class="header"
          data-testid="header-element"
        >
          <form
            class="form"
          >
            <input
              class="cityField"
              id="city"
              name="city"
              placeholder="placeholder"
              type="text"
            />
            <button
              class="addBtn"
            >
              buttons.add
            </button>
          </form>
          <button
            class="removeAll"
          >
            buttons.remove
          </button>
          <select
            class="selectLang"
            id="lang"
            name="language"
          >
            <option
              value="en"
            >
              EN
            </option>
            <option
              value="ru"
            >
              RU
            </option>
            <option
              value="uk"
            >
              UK
            </option>
            <option
              value="he"
            >
              HE
            </option>
          </select>
        </header>
        <h2
          class="text"
        >
          empty
        </h2>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="container-element"
    >
      <header
        class="header"
        data-testid="header-element"
      >
        <form
          class="form"
        >
          <input
            class="cityField"
            id="city"
            name="city"
            placeholder="placeholder"
            type="text"
          />
          <button
            class="addBtn"
          >
            buttons.add
          </button>
        </form>
        <button
          class="removeAll"
        >
          buttons.remove
        </button>
        <select
          class="selectLang"
          id="lang"
          name="language"
        >
          <option
            value="en"
          >
            EN
          </option>
          <option
            value="ru"
          >
            RU
          </option>
          <option
            value="uk"
          >
            UK
          </option>
          <option
            value="he"
          >
            HE
          </option>
        </select>
      </header>
      <h2
        class="text"
      >
        empty
      </h2>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
